libgdf/src/gdf_type_dispatcher.cuh

```diff
@@ -0,0 +1,95 @@
+ *
+ * auto example_lambda = [](auto dispatched_type_var){ 
+ *   using dispatched_type = decltype(dispatched_type_var);
+ *   return sizeof(dispatched_type);
+ * };
+ *
+ * gdf_type_dispatcher(GDF_INT8, example_lambda, 0);  // returns 1
+ * gdf_type_dispatcher(GDF_INT64, example_lambda, 0); // returns 8
+ *
+ * NOTE: A generic lambda can be thought of as if it were the following:
+ *
+ * struct example_lambda_closure{
+ *  template <typename T>
+ *  int operator()(T dispatched_type_var){
+ *    return sizeof(T);
+ * };
+ *
+ * Therefore, when using a generic lambda and the type_dispatcher, the lambda's first 
+ * parameter *must* be of type "auto" and the purpose of this parameter is solely
+ * to retrieve the dispatched type.
+ *
+ * @Param dtype The gdf_dtype enum that determines which type will be dispatched
+ * @Param f The functor with a templated "operator()" that will be invoked with 
+ * the dispatched type
+ * @Param args A parameter-pack (i.e., arbitrary number of arguments) that will 
+ * be perfectly-forwarded as the arguments of the functor's "operator()".
+ *
+ * @Returns Whatever is returned by the functor's "operator()". 
+ *
+ * NOTE: The return type for all template instantiations of the functor's "operator()" 
+ * must be the same, otherwise there will be a compiler error as you would be 
+ * trying to return different types from the same function.
+ */
+/* ----------------------------------------------------------------------------*/
+// This pragma disables a compiler warning that complains about the valid usage
+// of calling a __host__ functor from this function which is __host__ __device__
+#pragma hd_warning_disable
+template <class functor_t, typename... Ts>
+__host__ __device__ __forceinline__
+decltype(auto) gdf_type_dispatcher(gdf_dtype dtype, functor_t f, Ts&&... args)
+{
+    switch(dtype)
+    {
+      // The .template is known as a "template disambiguator" 
+      // See here for more information: https://stackoverflow.com/questions/3786360/confusing-template-error
+      case GDF_INT8:      { return f.template operator()<int8_t>(std::forward<Ts>(args)...); }
+      case GDF_INT16:     { return f.template operator()<int16_t>(std::forward<Ts>(args)...); }
+      case GDF_INT32:     { return f.template operator()<int32_t>(std::forward<Ts>(args)...); }
+      case GDF_INT64:     { return f.template operator()<int64_t>(std::forward<Ts>(args)...); }
+      case GDF_FLOAT32:   { return f.template operator()<float>(std::forward<Ts>(args)...); }
+      case GDF_FLOAT64:   { return f.template operator()<double>(std::forward<Ts>(args)...); }
+      case GDF_DATE32:    { return f.template operator()<int32_t>(std::forward<Ts>(args)...); }
+      case GDF_DATE64:    { return f.template operator()<int64_t>(std::forward<Ts>(args)...); }
+      case GDF_TIMESTAMP: { return f.template operator()<int64_t>(std::forward<Ts>(args)...); }
+      case GDF_CATEGORY:  { return f.template operator()<int32_t>(std::forward<Ts>(args)...); }
```

Comment:
I just realized there's a fundamental problem here.

No, this doesn't work. This is wrong. You see, there are different `gdf_type` values which use the _same_ type - and that means you can't actually replace the switch statement with a templated lambda or functor right now.

No, I'm sorry, you first have differentiate things type-wise. Either change the `types.h` interface to introduce appropriate wrappers (e.g. `typedef struct { int64_t value; } gdf_timestamp;` instead of `typedef int64_t gdf_timestamp`) - or create an implementaion-only templated tag type.

Reply:
@eyalroz It would help if you can provide the example of where you dispatch on `gdf_dtype` and you need to disambiguate more than the underlying data type? A specific link to code or sample code would help to get this right.

Reply:
@harrism : When working with GDF_TIMESTAMP's, you need to normalize by different time units; when working with GDF_DATE64's - you don't. Example: [gdf_extract_datetime_year()](https://github.com/rapidsai/cudf/blob/612ca60a7f41adb227ad556a91c76b69d2527d9b/libgdf/src/tests/datetime/datetimeops-test.cu#L108) for example.

Reply:
Better link: https://github.com/rapidsai/cudf/blob/612ca60a7f41adb227ad556a91c76b69d2527d9b/libgdf/src/datetimeops.cu#L312

That function doesn't dispatch on every gdf_dtype. And it is not dispatching in device code.